# **CASE STUDY 2 - Student Scoring Assessment**

> click **> Run All** in the toolbar above to run all cells at once


# 1. Overview



## **1.1 - Dataset Context**

Dataset: **AI-based student exam scoring log**

This dataset simulates the output of an automated exam scoring system used by a school. Each row represents one student and their score. The ai_score column is set to represent values generated by the AI system, not the actual student exam scores assessed manually. Dataset was generated by generative AI.



> *Preparation : Run the code below to load the dataset (shortcut: ctrl + Enter)*

In [1]:
import pandas as pd
import numpy as np
from IPython.display import display

# AI-based student exam scoring log
data = {
    "student_id": ["S001", "S002", "S003", "S004", "S005",
                   "S006", "S007", "S008", "S009", "S010",
                   "S011", "S012", "S013", "S014", "S015",
                   "S016", "S017", "S018"],
    "name": ["Jonathan", "Aisha", "Jonas", "Zara", "Liam",
             "Fatima", "Ahmed", "Sofia", "Kai", "Elena",
             "Mateo", "Amara", "Kenichi", "Leila", "Idris",
             "Anya", "Ryan", "Omar"],
    "gender": ["Male", "Female", "Male", "Female", "Male",
               "Female", "Male", "Female", "Male", "Female",
               "Male", "Female", "Male", "Female", "Male",
               "Female", "Male", "Male"],
    "class": ["A", "B", "A", "C", "B",
              "A", "C", "B", "C", "A",
              "A", "C", "B", "A", "C",
              "B", "B", "C"],
    "ai_score": [82, 78, None, 74, 88,
                 91, 76, 110, 83, 69,
                 95, 72, 80, -10, 85,
                 77, 88, 73],
    "status": ["pass", "pass", "pass", "fail", "pass",
               "pass", "fail", "pass", "pass", "fail",
               "Pass", "fail", "pass", "fail", "pass",
               "PASS", "pass", "fail"]
}

df = pd.DataFrame(data)

display(df)


,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass
1,S002,Aisha,Female,B,78.0,pass
2,S003,Jonas,Male,A,NaN,pass
3,S004,Zara,Female,C,74.0,fail
4,S005,Liam,Male,B,88.0,pass
5,S006,Fatima,Female,A,91.0,pass
6,S007,Ahmed,Male,C,76.0,fail
7,S008,Sofia,Female,B,110.0,pass
8,S009,Kai,Male,C,83.0,pass
9,S010,Elena,Female,A,69.0,fail


## **1.2 - Dataset Structure**

In [2]:
# Check dataset structure
print(f"Dataset header preview")
print('='*50)
display(df.head(1))


print(f"\n\nDataset shape (rows, columns)")
print('='*50)
display(df.shape)


print(f"\n\nDataset column types")
print('='*50)
display(df.dtypes)

Dataset header preview


,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass




Dataset shape (rows, columns)


(18, 6)



Dataset column types


,0
student_id,object
name,object
gender,object
class,object
ai_score,float64
status,object


From the output above, the Dataset Structure is as follows

-  **6** Headers; **student_id,	name,	gender,	class,	ai_score,	status**
-  **18** Rows, **6** Columns
-  Object data type; **student_id,	name,	gender,	class,	status**
-  Float64 data type (decimal); **ai_score**
-  Dataset type; **object**

> Dataset has been identified and audit can be carried out.

## **1.3 - Statistical and Categorical Summary**

In [13]:

# 1. Statistics
print(f"\nStatistical summary of the dataset")
print('='*50)
display(df.describe())

# 2. Categorical

## Find columns with non-standard data.
print(f"\n\nNumber of unique values per column")
print('='*50)
display(df.nunique())

## Found 'status' column has non-standard data.
print(f"\n\nLabel variants in 'status' column")
print('='*50)
display(np.unique(df['status'].dropna()))

## Find columns with NaN, None, or null values.
print(f"\n\nNumerical Columns which contain 'Nan, none, atau null'")
print('='*50)
display(df.isnull().sum())

## NaN found in numeric 'ai_score' column from the output above
print(f"\n\nRows with NaN, None, or null values")
print('='*50)
display(df[pd.to_numeric(df['ai_score'], errors='coerce').isnull()])







Statistical summary of the dataset


,ai_score
count,17.000000
mean,77.117647
std,24.583740
min,-10.000000
25%,74.000000
50%,80.000000
75%,88.000000
max,110.000000




Number of unique values per column


,0
student_id,18
name,18
gender,2
class,3
ai_score,16
status,4
status_normalized,2




Label variants in 'status' column


array(['PASS', 'Pass', 'fail', 'pass'], dtype=object)



Numerical Columns which contain 'Nan, none, atau null'


,0
student_id,0
name,0
gender,0
class,0
ai_score,1
status,0
status_normalized,0




Rows with NaN, None, or null values


,student_id,name,gender,class,ai_score,status,status_normalized
2,S003,Jonas,Male,A,NaN,pass,pass


From the output above, initial findings are
1. **Statistics**
 - Found values outside the allowed range, minimum value **min: -10** and maximum value **max: 110**. Violates the score scale rule of 100 (0 to 100).
- Standard Deviation is relatively large **std: 24** indicating high variation in student scores from the mean **mean: 77.11**.

2. **Categorical**
 - Columns **status** menghasilkan **4 label** yang seharusnya hanya **2 label** yaitu **pass** dan **fail**, likely due to inconsistent capitalization.
 - "Found **NaN** (Not a Number) value in the **ai_score** column that should be numeric, in student data for **Jonas** id **S003**




---

# 2. Analysis



## **2.1 - Audit Checklist**

1. Do students with scores outside the valid range get an appropriate graduation status?
2. Does the system have input validation to filter non-numeric values or values outside the 0 to 100 scoring scale?
3. Will inconsistent or non-standardized student status labels affect the system's decision output?
4. Is there any bias indication in the scoring system results toward student groups based on gender? Are the average scores and graduation rates distributed proportionally and fairly?

## **2.2 - Checklist 1 Solution**





**Checklist 1 - Step 1** : Finding values outside the 0-100 scale or non-valid values





In [4]:
# Filter students by 'ai_score' to find values outside the 0-100 scale or non-valid
df_invalid_scores = df[
    (df['ai_score'] > 100) |
    (df['ai_score'] < 0) |
    (df['ai_score'].isnull())
]
display(df_invalid_scores)

,student_id,name,gender,class,ai_score,status
2,S003,Jonas,Male,A,NaN,pass
7,S008,Sofia,Female,B,110.0,pass
13,S014,Leila,Female,A,-10.0,fail


### **Checklist 1 - Conclusion**

Do students with scores outside the valid range get an appropriate graduation status?
>No, because **3 students** were found with scores outside the 0-100 scale or invalid values:
- **S003** with value **NaN**, the system has no value to make a decision, but still produces status **pass**
- **S008** with value **110** exceeding the maximum valid limit, the system still processes and produces **pass**
- **S014** with value **-10** below the minimum valid limit, the system still processes and produces **fail**

## **2.3 - Checklist 2 Solution**

**Checklist 2 - Step 1** : Counting students with scores outside the 0-100 scale or invalid values

In [5]:
# Count students with invalid values
invalid_count = df_invalid_scores.shape[0]
print(f"Number of students with invalid values: {invalid_count}")

Number of students with invalid values: 3


**Checklist 2 - Step 2** : Counting students with valid scores within the 0-100 scale

In [6]:
valid_count = df.shape[0] - invalid_count
print(f"Number of students with valid values: {valid_count}")

Number of students with valid values: 15


### **Checklist 2 - Conclusion**

Does the system have input validation to filter non-numeric values or values outside the 0 to 100 scoring scale?
>No, because **3 students** with scores outside the 0-100 scale or invalid values still received a **graduation status** (**pass** or **fail**). Without input validation, the system cannot distinguish data that is fit to process from data that is not, which means all decisions the system produces for invalid data cannot be trusted.


## **2.4 - Checklist 3 Solution**

**Checklist 3 - Step 1** : Testing label format in the 'status' column




In [7]:
# Count occurrences of each label in 'status' column
display(df['status'].value_counts())


,count
status,
pass,10
fail,6
Pass,1
PASS,1


**Checklist 3 - Step 2** : Counting occurrences of standard labels (assumed 'pass' and 'fail')

In [8]:
# Filter 'pass' labels that match the standard and count occurrences
df_pass_standard= df[df['status'] =='pass']
display(df_pass_standard)
print(f"Number of rows with standard 'pass' label: {df_pass_standard.shape[0]}\n")
# Filter 'fail' labels that match the standard and count occurrences
df_fail_standard= df[df['status'] =='fail']
display(df_fail_standard)
print(f"Number of rows with standard 'fail' label: {df_fail_standard.shape[0]}\n")

,student_id,name,gender,class,ai_score,status
0,S001,Jonathan,Male,A,82.0,pass
1,S002,Aisha,Female,B,78.0,pass
2,S003,Jonas,Male,A,NaN,pass
4,S005,Liam,Male,B,88.0,pass
5,S006,Fatima,Female,A,91.0,pass
7,S008,Sofia,Female,B,110.0,pass
8,S009,Kai,Male,C,83.0,pass
12,S013,Kenichi,Male,B,80.0,pass
14,S015,Idris,Male,C,85.0,pass
16,S017,Ryan,Male,B,88.0,pass


Number of rows with standard 'pass' label: 10



,student_id,name,gender,class,ai_score,status
3,S004,Zara,Female,C,74.0,fail
6,S007,Ahmed,Male,C,76.0,fail
9,S010,Elena,Female,A,69.0,fail
11,S012,Amara,Female,C,72.0,fail
13,S014,Leila,Female,A,-10.0,fail
17,S018,Omar,Male,C,73.0,fail


Number of rows with standard 'fail' label: 6



### **Checklist 3 - Conclusion**

Will inconsistent or non-standardized student status labels affect the system's decision output?
>Yes, because the **status** column was found to have **4 categories**: **PASS, Pass, pass**, and **fail**. I assume the standard categories are **pass** and **fail** as the graduation benchmark:
- **10 students** with status **pass** passed
- **6 students** with status **fail** did not pass
- **2 students** have non-standard status labels (**1 student** with status **PASS** and **1 student** with status **Pass**)

Based on the output of <code><b>df_pass_standard = df[df['status'] == 'pass']</b></code> in step 2, students with status **PASS** and **Pass** are treated as not passing, because they do not match the system's standard.

So **2 students** who should have passed will be ignored by the system, because their status labels do not match the standard.




## **2.5 - Checklist 4 Solution**

**Checklist 4 - Step 1** : Normalizing non-standard formats for inclusion in the analysis
> *This step is done only for the purpose of the following analysis, not as a fix to the system. Label inconsistency remains a valid finding in checklist 3.*

In [9]:
# Create new column status_normalized to standardize graduation labels
df['status_normalized'] = df['status'].str.lower()
display(df)



,student_id,name,gender,class,ai_score,status,status_normalized
0,S001,Jonathan,Male,A,82.0,pass,pass
1,S002,Aisha,Female,B,78.0,pass,pass
2,S003,Jonas,Male,A,NaN,pass,pass
3,S004,Zara,Female,C,74.0,fail,fail
4,S005,Liam,Male,B,88.0,pass,pass
5,S006,Fatima,Female,A,91.0,pass,pass
6,S007,Ahmed,Male,C,76.0,fail,fail
7,S008,Sofia,Female,B,110.0,pass,pass
8,S009,Kai,Male,C,83.0,pass,pass
9,S010,Elena,Female,A,69.0,fail,fail


**Checklist 4 - Step 2** : Finding average 'ai_score' by 'gender'




In [10]:
# Find average 'ai_score' grouped by 'gender' column
display(df.groupby('gender')['ai_score'].agg(count_students='size', avg_ai_score='mean').reset_index())


,gender,count_students,avg_ai_score
0,Female,8,70.125000
1,Male,10,83.333333


**Checklist 4 - Step 3** : Finding graduation rate by 'gender' percentage.



In [11]:
# Find occurrence count of graduation labels 'status_normalized' grouped by 'gender' column
df_status_percentage = df.groupby('gender')['status_normalized'].value_counts().reset_index()

# Create 'percentage' column from occurrence count
df_status_percentage['percentage'] = (
    df_status_percentage['count'] /
    df_status_percentage.groupby('gender')['count'].transform('sum') * 100
)

display(df_status_percentage)



,gender,status_normalized,count,percentage
0,Female,fail,4,50.0
1,Female,pass,4,50.0
2,Male,pass,8,80.0
3,Male,fail,2,20.0


### **Checklist 4 - Conclusion**

Is there any indication of bias in the system's scoring results against student groups based on gender? Are average scores and graduation rates distributed proportionally and fairly?
> There is an indication of gender bias affecting graduation results that needs further investigation;
- The number of **female** students is **8** compared to **male** students at **10**, there is a proportional imbalance in the dataset that affects fairness.
- The average score of **female** students is lower than **male** students, at **70.12** compared to **83.33**. This difference also affects the system's assessment.
- **Female** students have a graduation rate of **pass 50%** vs **fail 50%**, while **male** students have **pass 80%** vs **fail 20%**. Further investigation is needed on the graduation ratio difference by gender to determine if this is systemic bias.

# 3. Audit Report



## **3.1 - Findings**

1. **3 rows** with values **NaN, -10,** and **110** still received a graduation status, even though those values are outside the 0-100 scoring scale. The system risks producing graduation decisions for invalid inputs because there is no input validation. Risk : **High**
2. The **status** column was found to have **4 categories**: **PASS, Pass, pass**, and **fail**. Assuming the standard categories are **pass** and **fail**, students with non-standard categories will be ignored by the system. Risk : **High**
3. The dataset shows high score variation (standard deviation **std: 24**). Investigation by gender shows an uneven gap;
>- The average score of **female** students at **70.12** is lower than **male** students at **83.33**.
>- A gap was also found in graduation rates, **female** students at **pass 50%** compared to **male** students at **pass 80%**.

>- The differences above need further investigation to determine if this reflects actual student ability or indicates systemic bias. Risk : **Medium**




## **3.2 - Recommendations**

1. Input validation needs to be established as a standard requirement so that values in the 'ai_score' column are always valid numbers within the 0-100 scale. This is directed to Developer/Engineer to follow up.
2. Standardization of categories in the status column needs to be applied so that no data is at risk of being ignored by the system due to non-compliance, even when students meet the graduation criteria. This needs to be handled by Developer/Engineer.
3. The data sample for average comparison per gender needs to be balanced in terms of gender representation. The graduation rate per gender is also higher for male students than female students. Further investigation is needed to determine whether the gender distribution difference in the dataset reflects bias in data collection or systemic bias in the model. Directed to the Data Engineer/Data Scientist/AI Ethics Analyst to evaluate the bias in the model.



## **3.3 - Limitations**

1. This dataset contains only 18 rows and was generated with the help of generative AI for audit practice purposes, findings need to be verified with a larger dataset before any conclusion can be drawn.
2. The fairness analysis in checklist 4 uses normalized labels, but the actual system ignores non-standard labels. Graduation rate percentages per gender may differ if calculated based on the actual system behavior.
3. Fairness audit was only conducted based on gender category. The class category was not analyzed because the sample size per class is too small to produce statistically significant conclusions.
4. Average ai_score per gender was calculated with NaN values ignored by default in Pandas, so the number of students counted per gender may not be equal.
5. This dataset does not include documentation on how ai_score was generated, whether it is purely model output, a transformation of original scores, or a combination of both. Audit conclusions need to be verified with complete system documentation.


